### PR TITLE
Bootstrap v4

### PR DESCRIPTION
**Pull request by antpaw**
This branch can be used for all the changes that will come with the update to v4

### CODE REVIEW

**Comment by mattbrictson** [comment]
Thanks for contributing this! Is this PR a work in progress, or is it ready for review?

If it is not ready yet, could you add "WIP" to the PR name?

**Comment by antpaw** [comment]
Hi, I'm building a rails CRUD app with `bootstrap_form` from my repo and `'bootstrap', '~> 4.0.0.alpha4'` and I can't detect any problems that would prevent from merging. However I wouldn't release a full major version of this, only a beta, because v4 is in alpha now.

**Comment by desheikh** [comment]
@antpaw ditto, looks like this works already in the bootstrap-v4 branch

**Comment by bradlis7** [comment]
Personally, I feel that since the bootstrap gem is keeping a v4 alpha up to date, it would be useful if we could get this as a gem as well (maybe require specifying the version if possible). Can we merge the changes from master into this branch as well?

**Comment by antpaw** [comment]
ok, just did it

**Comment by bradlis7** [comment]
Here's how you can use this branch in a project:

    // Gemfile
    gem 'bootstrap_form', git: 'https://github.com/bootstrap-ruby/rails-bootstrap-forms.git', branch: 'bootstrap-v4'

**Comment by antpaw** [comment]
@bradlis7 this branch is not my branch, they have started their own for some reason.

**Comment by bradlis7** [comment]
Well, that branch is somehow associated to this PR, as it says "View #299" when I open that branch. I assumed they pulled some of your changes in. It seems like there are about 5 or 6 people with changes for the conversion, so it's a little confusing.

**Comment by mattbrictson** [comment]
Hi, this repo has an official `bootstrap-v4` branch that we are using to consolidate all v4-related PRs. This is so we can review and merge improvements to v4 support incrementally without disrupting the master branch.

However we don't yet have an official PR to go with the `bootstrap-v4` branch, hence why people are getting directed here. I will create an official PR for v4 to clear up the confusion.

**Comment by mattbrictson** [comment]
Here is the official bootstrap-v4 PR: https://github.com/bootstrap-ruby/rails-bootstrap-forms/pull/315

@antpaw Have all your commits already been included in #315? If so can we close this PR?

**Comment by antpaw** [comment]
@mattbrictson ye they both look similar, it would be easier to review if your v4 branch also had the newest master in it..

**Comment by mattbrictson** [comment]
@antpaw good point; I have now rebased it on the latest `master`.

**Comment by antpaw** [comment]
hmm not sure why the diff is still so big, https://github.com/bootstrap-ruby/rails-bootstrap-forms/pull/299/files I like my version better e.g. the <p> tag for form errors and the def error_class

**Comment by mattbrictson** [comment]
@desheikh and @donv, since you've helped review v4 PRs before, would either of you mind taking a look at this one?

@antpaw Thanks for the PR and for changing the merge target to `bootstrap-ruby:bootstrap-v4`. Could you resolve the conflicts so it is easier to review?

**Comment by desheikh** [comment]
@mattbrictson @antpaw I think the reason the diff is so large is because this branch needs to be rebased against the bootstrap-v4 branch. Most of these changes already exist on bootstrap-v4.

**Comment by antpaw** [comment]
hmm it looks like github is bugged, the target branch is `bootstrap-ruby:bootstrap-v4` but i still see diffs like this one https://abload.de/img/screenshot2017-02-17a4rrb2.png but this should not have any diff if you look at the current state of the two branches

https://github.com/bootstrap-ruby/rails-bootstrap-forms/blob/bootstrap-v4/lib/bootstrap_form/form_builder.rb#L268
https://github.com/antpaw/rails-bootstrap-forms/blob/master/lib/bootstrap_form/form_builder.rb#L268

i will open an new pull request

**Comment by antpaw** [comment]
okay i have merged v4 into my v4

**Comment by antpaw** [comment]
one of the main changes is now that i have switched to a <p> tag for inline error and to a <div> for help text. this is done because this is how the documentation requires it. https://v4-alpha.getbootstrap.com/components/forms/?#help-text <span> tag is not a block element. The help text is not <p> because if you stack two <p>s (help and error) on top of each other it looks bad because of the padding.

File: lib/bootstrap_form/form_builder.rb [hunk 1]
```diff
@@ -203,6 +203,7 @@ def form_group(*args, &block)
         label = generate_label(options[:id], name, options[:label], options[:label_col], options[:layout]) if options[:label]
         control = capture(&block).to_s
         control.concat(generate_help(name, options[:help]).to_s)
+        # TODO create `generate_error`
```
Comment (desheikh) [inline lib/bootstrap_form/form_builder.rb]: remove todo tag

**Comment by desheikh** [comment]
@mattbrictson @antpaw need to remove the #TODO comment, looks good otherwise!

**Comment by antpaw** [comment]
done

**Comment by antpaw** [comment]
please note my other pr that covers this todo https://github.com/bootstrap-ruby/rails-bootstrap-forms/pull/300

**Comment by desheikh** [comment]
@antpaw @mattbrictson closing this since the changes in beta3 make this obsolete.
https://github.com/bootstrap-ruby/rails-bootstrap-forms/pull/370 covers the new syntax.

**Pull request closed without merging**